File: 리트코드/0148-sort-list/0148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head ==null || head.next == null) {
            return head;
        }

        // 러너 기법으로 중간 노드 탐색
        ListNode half = null, slow = head, fast = head;
        while (fast != null && fast.next != null) {
            half = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        // 중간 slow 기점으로 앞 뒤 연걸 끊기
        half.next = null;

        // 계속해서 좌우로 분할
        ListNode l1 = sortList(head);
        ListNode l2 = sortList(slow);

        return mergeTwoLists(l1, l2);
    }

    private ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }
        if (l2 == null) {
            return l1;
        }
        if (l1.val > l2.val) {
            ListNode tmp = l1;
            l1 = l2;
            l2 = tmp;
        }
        l1.next = mergeTwoLists(l1.next, l2);

        return l1;
    }
}

File: 리트코드/0148-sort-list/0148-sort-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun sortList(head: ListNode?): ListNode? {
        fun mergeTwoLists(l1: ListNode?, l2: ListNode?): ListNode? {
            if (l1 == null) {
                return l2
            }
            if (l2 == null) {
                return l1
            }

            return if (l1.`val` < l2.`val`) {
                l1.next = mergeTwoLists(l1.next, l2)
                l1
            } else {
                l2.next = mergeTwoLists(l1, l2.next)
                l2
            }
        }

        if(head?.next == null){
            return head
        }
        var half:ListNode? = null
        var slow:ListNode? = head
        var fast:ListNode? = head
        while (fast?.next != null) {
            half = slow
            slow = slow!!.next
            fast = fast.next.next
        }

        half!!.next = null

        var l1: ListNode? = sortList(head)
        var l2: ListNode? = sortList(slow)

        return mergeTwoLists(l1, l2)

    }
}